Objective-C/CBLProperties.mm

```diff
@@ -146,97 +145,403 @@ - (BOOL) containsObjectForKey: (NSString*)key {
 
 
 - (nullable id) objectForKeyedSubscript: (NSString*)key {
-    if (_properties)
-        return _properties[key];
-    else
-        return [self fleeceValueToObject: [self fleeceValueForKey: key]];
+    id obj = _properties ? _properties[key] : nil;
+    if (!obj) {
+        auto value = [self fleeceValueForKey: key];
+        obj = [self fleeceValueToObject: value  forKey: key asJson: NO];
+        [self cacheValue: obj forKey: key changed: NO];
+    }
+    return obj;
 }
 
 
 - (void) setObject: (nullable id)value forKeyedSubscript: (NSString*)key {
-    [self mutateProperties];
-    
     // NSDate:
     if ([value isKindOfClass: NSDate.class])
         value = [CBLJSON JSONObjectWithDate: value];
     
     id oldValue = _properties[key];
     
     if (![value isEqual: oldValue] && value != oldValue) {
```

Comment:
Since `_properties` is a cache, this method will do the wrong thing if called with a nil `value`, in a document where the key has a value but it hasn't been accessed yet. `oldValue` will be nil, so the `if` block will be skipped and nothing will change.

I think you need to change `_properties[key]` to `self[key]`.